File: python/encircle_public_api_client.py
#!/usr/bin/env python

from google.protobuf.wrappers_pb2 import StringValue, DoubleValue, UInt64Value
import grpc

import os
import os.path
import subprocess
import sys

def generate_bindings():
    script_dir = os.path.dirname(os.path.realpath(__file__))
    process = subprocess.Popen(
        (
            'python',
            '-m'
            'grpc_tools.protoc',
            '-I..',
            '--python_out=.',
            '--grpc_python_out=.',
            '../encircle_public_api_v1.proto',
        ),
        cwd=script_dir,
    )
    process.wait()

def main(ca_cert, client_cert, client_key, server_uri='api.encircleapp.com:50051'):
    from encircle_public_api_v1_pb2_grpc import PublicAPIV1Stub
    from encircle_public_api_v1_pb2 import (
        BrandFilterCriteria,
        NewPropertyClaimInfo,
        OrganizationFilterCriteria,
        PropertyClaimDocumentFilterCriteria,
        PropertyClaimFilterCriteria,
        PropertyClaimId,
        PropertyClaimPictureFilterCriteria,
        SymbolicDate,
    )

    with open(ca_cert, 'rb') as ca_cert_file,\
            open(client_cert, 'rb') as client_cert_file,\
            open(client_key, 'rb') as client_key_file:

        client_credentials = grpc.ssl_channel_credentials(
            root_certificates=ca_cert_file.read(),
            certificate_chain=client_cert_file.read(),
            private_key=client_key_file.read(),
        )

    channel = grpc.secure_channel(server_uri, client_credentials)

    stub = PublicAPIV1Stub(channel)

    # Lists all organizations
    for organization in stub.ListOrganizations(OrganizationFilterCriteria()):
        print("FOUND ORGANIZATION: {!r}".format(organization.id))

        # Lists all brands for a given organization
        for brand in stub.ListBrands(BrandFilterCriteria(organization_id=organization.id)):
            print("FOUND BRAND: {!r}".format(brand))

        # Lists all claims for a given organization
        for (i, claim) in enumerate(stub.ListPropertyClaims(PropertyClaimFilterCriteria(organization_id=organization.id))):
            print("FOUND CLAIM: {!r}".format(claim))

            # Lists all pictures in a given claim
            for p in stub.ListPropertyClaimPictures(PropertyClaimPictureFilterCriteria(claim_id=claim.id)):
                print("FOUND PICTURE: {!r}".format(p.download_uri))

            # Lists all rendered documents in a given claim
            for p in stub.ListPropertyClaimDocuments(PropertyClaimDocumentFilterCriteria(claim_id=claim.id)):
                print("FOUND DOCUMENT: {!r}".format(p.download_uri))

            if i >= 10:
                break

    # This example fills all fields, but most fields can be left out (and will either be blank or given a reasonable default).
    to_create = NewPropertyClaimInfo(
        organization_id=organization.id,
        brand_id=brand.id,
        insurer_identifier=StringValue(
            value="[PUBLIC API TEST CLAIM] INSURER IDENTIFIER",
        ),
        assignment_identifier=StringValue(
            value="[PUBLIC API TEST CLAIM] ASSIGNMENT IDENTIFIER",
        ),
        contractor_identifier=StringValue(
            value="[PUBLIC API TEST CLAIM] CONTRACTOR IDENTIFIER",
        ),
        policyholder_name=StringValue(
            value="[PUBLIC API TEST CLAIM] POLICYHOLDER NAME",
        ),
        loss_details=StringValue(
            value="LOSS DETAILS",
        ),
        full_address=StringValue(
            value="FULL ADDRESS",
        ),
        insurance_company_name=StringValue(
            value="INSURANCE COMPANY NAME",
        ),
        adjuster_name=StringValue(
            value="ADJUSTER NAME",
        ),
        broker_or_agent_name=StringValue(
            value="BROKER OR AGENT NAME",
        ),
        policyholder_email_address=StringValue(
            value="POLICYHOLDER EMAIL ADDRESS",
        ),
        policyholder_phone_number=StringValue(
            value="POLICYHOLDER PHONE NUMBER",
        ),
        policy_number=StringValue(
            value="POLICY NUMBER",
        ),
        date_of_loss=SymbolicDate(
            year=2000,
            month=SymbolicDate.June,
            day=1,
        ),
        date_claim_created=SymbolicDate(
            year=1999,
            month=SymbolicDate.August,
            day=15,
        ),
        default_depreciation=DoubleValue(
            value=0.1,
        ),
        max_depreciation=DoubleValue(
            value=0.9,
        ),
        sales_tax=DoubleValue(
            value=0.13,
        ),
        emergency_estimate=UInt64Value(
            value=150000,
        ),
        repair_estimate=UInt64Value(
            value=100000,
        ),
        contents_estimate=UInt64Value(
            value=5000,
        )
    )

    created = stub.CreatePropertyClaim(to_create)

    print("CREATED CLAIM: {}".format(created))

if __name__ == '__main__':
    generate_bindings()
    if len(sys.argv) not in (4, 5):
        print("USAGE: ./encircle_public_api_client.py FULL_PATH_TO_CA_CERT FULL_PATH_TO_CLIENT_CERT FULL_PATH_TO_CLIENT_KEY [SERVER_URI]")

    main(*sys.argv[1:])

File: php/encircle_public_api_client.php
<?php

$autoloader = require __DIR__ . '/vendor/autoload.php';

/**
 * @param string $ca_cert Path to Encircle CA certificate.
 *
 * @param string $client_cert Path to the client certificate. This is the
 * signed certificate generated for you based on the submitted CSR.
 *
 * @param string $client_key Path to the private key used to generate the CSR.
 * The previous 2 files consist of public information but this file should
 * never be disclosed to anyone, not even the Encircle team.
 *
 * @param string server_uri Hostname and port to connect to the API server.
 */
function main($ca_cert, $client_cert, $client_key, $server_uri) {
    $client = new Server\Public_api\PublicAPIV1Client($server_uri, [
        'credentials' => Grpc\ChannelCredentials::createSsl(
            file_get_contents($ca_cert),
            file_get_contents($client_key),
            file_get_contents($client_cert)
        ),
    ]);

    $org = getOrganization($client) or die("Could not find organization");
    $brand = getBrand($client, $org) or die("Could not find brand");

    print("Creating Claim in org {$org->getId()} with brand {$brand->getId()}\n");

    $claimInfo = new Server\Public_api\NewPropertyClaimInfo();

    // these fields are required
    $claimInfo->setOrganizationId($org->getId());
    $claimInfo->setBrandId($brand->getId());

    // one of these 3 fields should be set
    $claimInfo->setInsurerIdentifier(makeString("[PUBLIC API TEST CLAIM] INSURER IDENTIFIER"));
    $claimInfo->setAssignmentIdentifier(makeString("[PUBLIC API TEST CLAIM] ASSIGNMENT IDENTIFIER"));
    $claimInfo->setContractorIdentifier(makeString("[PUBLIC API TEST CLAIM] CONTRACTOR IDENTIFIER"));

    // all of these fields are optional and can be omitted
    $claimInfo->setPolicyholderName(makeString("[PUBLIC API TEST CLAIM] POLICYHOLDER NAME"));
    $claimInfo->setLossDetails(makeString("LOSS DETAILS"));
    $claimInfo->setFullAddress(makeString("FULL ADDRESS"));
    $claimInfo->setInsuranceCompanyName(makeString("INSURANCE COMPANY NAME"));
    $claimInfo->setAdjusterName(makeString("ADJUSTER NAME"));
    $claimInfo->setBrokerOrAgentName(makeString("BROKER OR AGENT NAME"));
    $claimInfo->setPolicyholderEmailAddress(makeString("POLICYHOLDER EMAIL ADDRESS"));
    $claimInfo->setPolicyholderPhoneNumber(makeString("POLICYHOLDER PHONE NUMBER"));
    $claimInfo->setPolicyNumber(makeString("POLICY NUMBER"));
    $claimInfo->setDateOfLoss(makeDate(2000, Server\Public_api\SymbolicDate_Month::June, 1));
    $claimInfo->setDateClaimCreated(makeDate(2000, Server\Public_api\SymbolicDate_Month::August, 15));
    $claimInfo->setDefaultDepreciation(makeDouble(0.1));
    $claimInfo->setMaxDepreciation(makeDouble(0.9));
    $claimInfo->setSalesTax(makeDouble(0.13));
    $claimInfo->setEmergencyEstimate(makeInt(150000));
    $claimInfo->setRepairEstimate(makeInt(100000));
    $claimInfo->setContentsEstimate(makeInt(5000));

    list($claimResult, $status) = $client->CreatePropertyClaim($claimInfo)->wait();
    if ($status->code != Grpc\STATUS_OK) {
        die("Could not create claim, error={$status->code}, details={$status->details}");
    }

    print("Created Claim with ID = {$claimResult->getId()}\n");
}

function getOrganization($client) {
    $orgFilter = new Server\Public_api\OrganizationFilterCriteria();
    $orgResponses = $client->ListOrganizations($orgFilter)->responses();
    foreach ($orgResponses as $org) {
        return $org;
    }
    return null;
}

function getBrand($client, $org) {
    $brandFilter = new Server\Public_api\BrandFilterCriteria();
    $brandFilter->setOrganizationId($org->getId());
    $brandResponses = $client->ListBrands($brandFilter)->responses();
    foreach ($brandResponses as $brand) {
        return $brand;
    }
    return null;
}

function makeString($str) {
    $stringValue = new Google\Protobuf\StringValue();
    $stringValue->setValue($str);
    return $stringValue;
}

function makeDate($year, $month, $day) {
    $date = new Server\Public_api\SymbolicDate();
    $date->setYear($year);
    $date->setMonth($month);
    $date->setDay($day);
    return $date;
}

function makeDouble($num) {
    $doubleValue = new Google\Protobuf\DoubleValue();
    $doubleValue->setValue($num);
    return $doubleValue;
}

function makeInt($num) {
    $intValue = new Google\Protobuf\UInt64Value();
    $intValue->setValue($num);
    return $intValue;
}

if (count($argv) < 4) {
    $usage = <<<HERE
Usage: php {$argv[0]} CA_CERT CLIENT_CERT CLIENT_KEY [SERVER_URI]

CA_CERT is the full path to the Encircle CA certificate.

CLIENT_CERT is the full path to your client certificate. This is the signed
certificate generated for you based on the submitted CSR.

CLIENT_KEY is the full path to your client PRIVATE KEY. The previous 2 files
consist of public information but the contents of this file should be kept
secret and not shared with anyone!

SERVER_URI points to `api.encircleapp.com:50051` by default but can be used to
point to a different API server.

HERE;

    die($usage);
} else {
    main(
        $argv[1],
        $argv[2],
        $argv[3],
        array_key_exists(4, $argv) ? $argv[4] : 'api.encircleapp.com:50051'
    );
}
